sktime/forecasting/statsforecast.py

```diff
@@ -552,3 +555,150 @@ def get_test_params(cls, parameter_set="default"):
+        not supported.
+
+    References
+    ----------
+    .. [1]
+        https://nixtla.github.io/statsforecast/src/core/models.html#multiple-seasonalities
+
+    Examples
+    --------
+    >>> from sktime.datasets import load_airline
+    >>> from sktime.forecasting.statsforecast import StatsForecastMSTL
+
+    >>> y = load_airline()
+    >>> model = StatsForecastMSTL(season_length=[3,12]) # doctest: +SKIP
+    >>> fitted_model = model.fit(y=y) # doctest: +SKIP
+    >>> y_pred = fitted_model.predict(fh=[1,2,3]) # doctest: +SKIP
+    """
+
+    _tags = {
+        "ignores-exogeneous-X": True,
+        "capability:pred_int": True,
+        "capability:pred_int:insample": True,
+        "python_dependencies": ["statsmodels", "statsforecast"],
+    }
+
+    def __init__(
+        self,
+        season_length: Union[int, List[int]],
+        trend_forecaster=None,
+    ):
+        super().__init__()
+
+        if _check_soft_dependencies("statsmodels", severity="none"):
+            from statsforecast.models import _TS
+        else:
+            raise Exception(
+                "StatsForecastMSTL requires module `statsforecast` to be installed. "
+                "Please install `statsforecast` in your environment."
+            )
+
+        self.trend_forecaster = trend_forecaster
+        self.season_length = season_length
+        if trend_forecaster:
+            self._trend_forecaster = trend_forecaster
+        else:
+            self._trend_forecaster = StatsForecastAutoETS(model="ZZN")
+
+        # checks if trend_forecaster is already wrapped with
+        # StatsForecastBackAdapterMSTL
+        if not isinstance(self._trend_forecaster, StatsForecastBackAdapter):
+            # if trend_forecaster is sktime forecaster
+            if isinstance(self._trend_forecaster, BaseForecaster):
+                self._trend_forecaster = StatsForecastBackAdapter(
+                    self._trend_forecaster
+                )
+            # if trend_forecaster is not StatsForecast forecaster
+            elif not isinstance(self._trend_forecaster, _TS):
+                raise Exception(
+                    "The provided forecaster is not compatible with MSTL. Please ensure"
+                    " that the forecaster you pass into the model is a sktime "
+                    "forecaster or statsforecast forecaster."
+                )
+
+    def _instantiate_model(self):
+        """Create underlying forecaster instance."""
+        from statsforecast.models import MSTL
+
+        return MSTL(
+            season_length=self.season_length,
+            trend_forecaster=self._trend_forecaster,
+        )
+
+    def __getattr__(self, name):
+        """Return self._trend_forecaster when self.trend_forecaster is accessed."""
+        if name == "trend_forecaster":
+            return self._trend_forecaster
+        else:
+            return self.__getattribute__(name)
```

Comment:
If this is the behaviour, why not assign as such in the `__init__` directly? I'm not following the logic, can you please explain?

Reply:
I did this because fitting `self.trend_forecaster` would be overwriting class params, which afaik is not allowed because of a contract

Reply:
@yarnabrina is right, I think this is dangerous to do, and will break the `set_params` / `get_params` logic! The tests would tell you if `get_test_params` would return two parameter sets (I'd recommend, add a second one so all `set_params` / `get_params` tests run).

The "standard protocol" would be:

* in `__init__`, write `trend_forecaster` to `self.trend_forecaster`
* in `__init__` (or anytime before the logic in `_fit`), write a clone of `trend_forecaster` to `self._trend_forecaster`
* if you need a default, e.g., for `trend_forecaster = None`, write the default to `self._trend_forecaster` in the `None` case
* never access `self.trend_forecaster` after this
* whenever you need the estimator that needs to mutate state, e.g., by fitting, use `self._trend_forecaster`

Reply:
Gotcha, thanks guys!